### PR TITLE
Fix OpenTelemetrySdk(.tracerProvider).tracerBuilder() being noop.

### PR DESCRIPTION
**Pull request by Oberon00**
The default method implementation of TracerProvider.tracerBuilder and OpenTelemetry.tracerBuilder is not very helpful, as demonstrated by this PR.

~~I will amend the PR later to provide a fix, just wanted to give you a quick heads-up  since the release of 1.5 seems imminent -- probably this should be fixed before. @jkwatson @anuraaga~~

### CODE REVIEW

**Review by anuraaga — APPROVED** [review]
Thanks @Oberon00 - agree the default implementation was probably a miss, but glad we're able to work with it without breaking compatibility

**Comment by Oberon00** [comment]
I think an alternative default implementation would be to return a working tracer provider that uses the other interface methods and just ignores setSchemaUrl.

**Comment by Oberon00** [comment]
Build `(macos-latest, 17)` failed with a probably unrelated error:

```
> Task :sdk:trace:test

BatchSpanProcessorTest > exporterThrowsException() FAILED
    java.lang.NullPointerException: Cannot invoke "io.opentelemetry.sdk.common.CompletableResultCode.whenComplete(java.lang.Runnable)" because "code" is null
        at io.opentelemetry.sdk.common.CompletableResultCode.ofAll(CompletableResultCode.java:48)
        at io.opentelemetry.sdk.trace.export.MultiSpanExporter.shutdown(MultiSpanExporter.java:94)
        at io.opentelemetry.sdk.trace.export.BatchSpanProcessor$Worker.lambda$shutdown$2(BatchSpanProcessor.java:268)
        at io.opentelemetry.sdk.common.CompletableResultCode.whenComplete(CompletableResultCode.java:126)
        at io.opentelemetry.sdk.trace.export.BatchSpanProcessor$Worker.shutdown(BatchSpanProcessor.java:265)
        at io.opentelemetry.sdk.trace.export.BatchSpanProcessor$Worker.access$200(BatchSpanProcessor.java:125)
        at io.opentelemetry.sdk.trace.export.BatchSpanProcessor.shutdown(BatchSpanProcessor.java:110)
        at io.opentelemetry.sdk.trace.TracerSharedState.shutdown(TracerSharedState.java:96)
        at io.opentelemetry.sdk.trace.SdkTracerProvider.shutdown(SdkTracerProvider.java:115)
        at io.opentelemetry.sdk.trace.export.BatchSpanProcessorTest.cleanup(BatchSpanProcessorTest.java:69)
```

That looks more serious than a simple timeout. Should I create an issue?

File: sdk/all/src/main/java/io/opentelemetry/sdk/OpenTelemetrySdk.java [hunk 1]
```diff
@@ -73,6 +74,11 @@ public Tracer get(String instrumentationName, String instrumentationVersion) {
       return delegate.get(instrumentationName, instrumentationVersion);
     }
 
+    @Override
+    public TracerBuilder tracerBuilder(String instrumentationName) {
+      return delegate.tracerBuilder(instrumentationName);
+    }
```
Comment (jkwatson) [inline sdk/all/src/main/java/io/opentelemetry/sdk/OpenTelemetrySdk.java:80]: why is this necessary? The default implementation should cover this, precisely, no?

File: api/all/src/main/java/io/opentelemetry/api/GlobalOpenTelemetry.java [hunk 2]
```diff
@@ -206,5 +206,10 @@ public TracerProvider getTracerProvider() {
     public ContextPropagators getPropagators() {
       return delegate.getPropagators();
     }
+
+    @Override
+    public TracerBuilder tracerBuilder(String instrumentationName) {
+      return delegate.tracerBuilder(instrumentationName);
+    }
```
Comment (jkwatson) [inline api/all/src/main/java/io/opentelemetry/api/GlobalOpenTelemetry.java:213]: do I completely not understand how default methods and method dispatch works or something? I don't see why this is necessary given the default implementation.

File: sdk/all/src/main/java/io/opentelemetry/sdk/OpenTelemetrySdk.java (hunk 1 above)
Comment (Oberon00) [reply sdk/all/src/main/java/io/opentelemetry/sdk/OpenTelemetrySdk.java:80]: `ObfuscatedTracerProvider` derives from the `TracerProvider` interface which has this default impl:

https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/trace/TracerProvider.java#L58-L60

`return DefaultTracerBuilder.getInstance();` is a no-op TracerBuilder.

Comment (jkwatson) [reply sdk/all/src/main/java/io/opentelemetry/sdk/OpenTelemetrySdk.java:80]: ah. the issue here is in the ObfuscatedTracerProvider, not in this class at all. We should definitely fix the ObfuscatedTracerProvider. This particular class I don't care so much about since it works as its supposed to.

File: api/all/src/main/java/io/opentelemetry/api/GlobalOpenTelemetry.java (hunk 2 above)
Comment (Oberon00) [reply api/all/src/main/java/io/opentelemetry/api/GlobalOpenTelemetry.java:213]: Here, the default implementation actually does something sensible:

https://github.com/open-telemetry/opentelemetry-java/blob/d9e1151336afc5778a110863bb9aa4178af567e3/api/all/src/main/java/io/opentelemetry/api/trace/TracerProvider.java#L75-L77

Still, we bypass any override that the delegate might have, calling `delegate.getTracerProvider().tracerBuilder(instrumentationName)` instead of `delegate.tracerBuilder(instrumentationName)`. This will hopefully do the same, but this bypass would still be very surprising in case one adds logging to `tracerBuilder` or an auto-instrumenting agent instruments it, etc.

File: sdk/all/src/main/java/io/opentelemetry/sdk/OpenTelemetrySdk.java (hunk 1 above)
Comment (Oberon00) [reply sdk/all/src/main/java/io/opentelemetry/sdk/OpenTelemetrySdk.java:80]: This *is* the ObfuscatedTracerProvider which is a static inner class of OpenTelemetrySdk.

File: api/all/src/main/java/io/opentelemetry/api/GlobalOpenTelemetry.java (hunk 2 above)
Comment (jkwatson) [reply api/all/src/main/java/io/opentelemetry/api/GlobalOpenTelemetry.java:213]: I assume this is also due to the missing implementation in the ObfuscatedTracerProvider. I'd rather fix it there, rather than patch the issue and not fix the underlying problem.

Comment (jkwatson) [reply api/all/src/main/java/io/opentelemetry/api/GlobalOpenTelemetry.java:213]: ugh. I also missed that this was the implementation in the obfuscated impl. Sorry for the noise.

**Review by jkwatson — APPROVED** [review]
thanks!
